00XX-efficient-msms-for-non-native-pickles-verification.md

```diff
@@ -0,0 +1,481 @@
+
+
+Regarding ECC, o1js also contains an implementation that is part of ECDSA library. It is unclear though whether this can be useful in the kimchi context, as this library is written in o1js/typescript --- probably it can be used as an inspiration.
+- https://github.com/o1-labs/o1js/blob/main/src/lib/gadgets/elliptic-curve.ts#L99
+
+Additionally, one will need to decide which coordinates would are better to use. Available options are:
+- [Affine](https://www.hyperelliptic.org/EFD/g1p/auto-shortw.html)
+- [Jacobian](https://www.hyperelliptic.org/EFD/g1p/auto-shortw-jacobian.html#doubling-dbl-2007-bl)
+- [Projective](https://www.hyperelliptic.org/EFD/g1p/auto-shortw-projective.html)
+
+Again, **in practice** affine coordinates will be most likely the most efficient, and also easy to implement, so the suggested path is to start by implementing those, and revisiting the technique much later in the project timeline.
+
+
+We might find inspiration and decide which encoding is more efficient by looking at existing FF/ECC implementations:
+- https://github.com/privacy-scaling-explorations/halo2wrong
+- https://github.com/DelphinusLab/halo2ecc-s
+- https://hackmd.io/@arielg/B13JoihA8
+
+
+### Algorithm Performance, Circuit Layout, and Multi-Circuit Folding
+
+The circuit that we are implementing is large, and its layout should be carefully considered. Luckily, we have some liberty in design decisions -- the algorithm needs to be implemented on a *variant* of Kimchi which can allow long rows, additive lookups (memory access), and support for folding. Two last features are (being) developed in the zkVM project.
+
+Note that each iteration (one run) of the sub-MSM algorithm with limited buckets takes $3n$ additions at most:
+- The cycle in the beginning populates the buckets and it takes $n$ additions because that's how much multiplications are there in `to_scale_pairs`.
+- The end loop does two additions per each iteration (and each iteration assumes a non-zero bucket, assume there are $n'$ of these buckets), so we have $2 \cdot n'$ additions. Since $n' < n$, the worst case is $2 n$ additions.
+
+In addition to $3n$ addition per sub-MSM run, we will need to sum all the results of it. Luckily, this computation is relatively cheap --- we require one extra addition per run.
+
+In total, the whole algorithm then requires $3n \cdot l$ additions because we need to run the inner algorithm $l$ times, and then sum the results (which is negligibly small and can be accumulated along the way). Assuming worst (of the two) case of $n = 2^{16}$ and $k = 15$, we get about $2^{17}$ additions, which is still more than $2^{16}$ budget. But even with the easier $n = 2^{15}$ we still have to fit $3$ times more FF EC additions than there are rows available.
```

Comment:
See above. What we actually want is `l * n + 2*n`. What you're describing is roughly 2.5x worse that the original design.

Reply:
Fair -- misread the size of the array, will have to update the estimates paragraph.